File: src/medium/sort-array.ts
export const sortArray = (nums: number[]): number[] => {
	const flat = nums.flat(Infinity);
	if (flat.length === 0) return [];
	const freq = new Map<number, number>();
	const result = new Array<number>(flat.length);
	let max = flat[0];
	let min = flat[0];
	let index = 0;
	for (const num of flat) {
		freq.set(num, (freq.get(num) ?? 0) + 1);
		if (num > max) max = num;
		if (num < min) min = num;
	}
	for (let i = min; i <= max; i++) {
		let count = freq.get(i) ?? 0;
		while (count > 0) {
			count--;
			result[index++] = i;
		}
	}
	return result;
};

File: src/medium/sort_array.rs
use std::collections::HashMap;

pub fn sort_array(nums: Vec<i32>) -> Vec<i32> {
    let mut freq: HashMap<i32, i32> = HashMap::with_capacity(nums.len());
    if nums.is_empty() {
        return nums;
    }

    let mut result: Vec<i32> = vec![0; nums.len()];
    let mut max = nums[0];
    let mut min = nums[0];
    let mut index: usize = 0;
    for &num in &nums {
        freq.entry(num).and_modify(|x| *x += 1).or_insert(1);
        if max < num {
            max = num;
        }
        if min > num {
            min = num;
        }
    }
    for i in min..=max {
        if let Some(count) = freq.get_mut(&i) {
            while count > &mut 0 {
                *count -= 1;
                result[index] = i;
                index += 1;
            }
        }
    }
    result
}

